050-Kafka-Node-Pools.md

```diff
@@ -0,0 +1,832 @@
+In the KRaft mode, the Kafka cluster will consist only from Kafka nodes.
+But they will have different roles - they will be either controllers, or brokers, or will combine both of them.
+The controller role means the node is responsible for maintaining the cluster quorum, for the elections, discovery and for handling the cluster metadata instead of ZooKeeper.
+The broker role means the node is responsible for handling messages as a regular broker in the old ZooKeeper based clusters.
+And finally, the combined role means the node is responsible for both at the same time.
+
+The KRaft based Kafka clusters will be typically used in 3 different architectures:
+
+1. One or more nodes combining the controller and broker roles
+2. One or more nodes with the controller roles and one or more nodes with the broker roles
+3. One or more nodes with the combined role and one or more nodes with the broker roles
+
+In Strimzi, we want to support all of these architectures.
+But we also want to make it possible to transition between them.
+Since Apache Kafka plans to remove ZooKeeper support in its 4.0 release, supporting KRaft mode in its different form is not really an option for us but something we have to do.
+
+### Stretch clusters
+
+There seems to be increased demand Kafka clusters stretched across multiple Kubernetes clusters.
+Apache Kafka itself has latency limitations and might not run well when stretched over distant clusters with big latency.
+But there are many use cases where stretched clusters might be useful.
+For example:
+
+* When running closely colocated Kubernetes clusters
+* When migrating from old Kubernetes cluster to new Kubernetes cluster
+
+## Proposal
+
+This proposal suggests introduction of Kafka _node pools_.
+_Node pools_ are groups of Kafka nodes which will have the same configuration.
+A Kafka cluster can consist of one or more pools.
+Each node pool can belong to only one Kafka cluster.
+
+### `KafkaNodePool` custom resource
+
+A new custom resource called `KafkaNodePool` will be added to Strimzi.
+One `KafkaNodePool` resource will represent one _node pool_.
+It will take over some of the configuration from the `Kafka` CR, reuse other parts from it and add some new configuration options as well.
+The `KafkaNodePool` resource will be a _public_ resource.
+It will be created and managed directly by users (unlike for example the `StrimziPodSet` resource which is used only internally by Strimzi and should not be touched by the users).
+
+The _cluster-wide_ configuration options will remain part of the `Kafka` CR.
+While the _pool-specific_ options will be taken from `KafkaNodePool` resources
+
+For the initial implementation, the following options are proposed for the `KafkaNodePool` resource and its `.spec` section:
+
+* `replicas` indicating the number of nodes in the pool
+* `storage` indicating the storage of the nodes in this pool
```

Comment:
And what would be the actual benefit? It would be quite complex in terms of implementation. Today, the proposal is changing the 1:N relationship (one Kafka CR to multiple pods) to N:M relationship (multiple NodePools against multiple pods). Having a separate storage configuration would change it to N:M:O relationship (multiple NodePools against multiple pods against multiple storage pools). 

It would also further complicate the user experience as even a very basic cluster (3 KRaft controllers + 3 brokers) would likely require at least two storage pools (different size for controllers versus brokers) and two node pools (controllers and brokers).

I'm not sure I understand the advantages it would bring. WDYM with vertical scaling in this case? The node pool can be vertically scaled -> you can just change the resources and/or storage size without any limitations.

Reply:
I miss have miss understood, I had taken away the impression that the increase the resources available the node pool needed replaced as that's not the case it doesn't bring any advantage.I miss have miss understood, I had taken away the impression that the increase the resources available the node pool needed replaced as that's not the case it doesn't bring any advantage.